File: languages/python/04-expert/smart-pointers-in-depth/example/main.py
# Module focus: Tracking ownership and lifetime when multiple references can observe the same value.
# Why it matters: the example makes it possible to model exclusive, shared, and non-owning
# relationships idiomatically before the learner tackles the exercises.

# This example shows tracking ownership and lifetime when multiple references can observe the same
# value.

from __future__ import annotations

import gc
import weakref


# Separate helpers keep the main path focused on how to model exclusive, shared, and non-owning
# relationships idiomatically.
class Report:
    def __init__(self, title: str) -> None:
        # These values exercise the normal path before the exercises vary the documented
        # boundaries.
        self.title = title
        # The printed result shows whether the program can prevent leaks, cycles, and stale
        # observations in ownership graphs.
        print(f"Created report: {title}")


class ReportOwner:
    def __init__(self, name: str, report: Report | None) -> None:
        self._name = name
        self._report = report

    def transfer_to(self, destination: "ReportOwner") -> None:
        if self._report is None:
            print(f"{self._name} has nothing to transfer.")
            return
        print(f"{self._name} transfers {self._report.title} to {destination._name}.")
        destination._report = self._report
        self._report = None

    def describe(self) -> None:
        print(f"{self._name}: {'empty' if self._report is None else self._report.title}")


class PreviewPane:
    def __init__(self, report: Report) -> None:
        self._report_ref = weakref.ref(report)

    def describe(self) -> None:
        report = self._report_ref()
        if report is None:
            print("Preview target expired.")
            return
        print(f"Preview can still see: {report.title}")


class PreviewSession:
    def __init__(self, report: Report) -> None:
        self._report = report

    @property
    def current(self) -> Report | None:
        return self._report

    def release(self) -> None:
        self._report = None


def create_preview_session() -> tuple[PreviewPane, PreviewSession]:
    session = PreviewSession(Report("Transient Draft"))
    current = session.current
    assert current is not None
    return PreviewPane(current), session


# Fixed inputs make the consequence of assuming a weak reference behaves like a normal strong
# reference visible and repeatable.
def main() -> None:
    inbox = ReportOwner("Inbox", Report("Quarterly Summary"))
    archive = ReportOwner("Archive", None)

    inbox.describe()
    archive.describe()
    inbox.transfer_to(archive)
    inbox.describe()
    archive.describe()

    preview, session = create_preview_session()
    preview.describe()

    session.release()
    gc.collect()
    preview.describe()


if __name__ == "__main__":
    main()

File: languages/go/04-expert/smart-pointers-in-depth/example/main.go
// Module focus: Tracking ownership and lifetime when multiple references can observe the same value.
// Why it matters: the example makes it possible to model exclusive, shared, and non-owning
// relationships idiomatically before the learner tackles the exercises.

package main

import "fmt"

// Separate helpers keep the main path focused on how to model exclusive, shared, and non-owning
// relationships idiomatically.
type report struct {
	title string
}

type reportOwner struct {
	name   string
	report *report
}

func (o *reportOwner) transferTo(destination *reportOwner) {
	if o.report == nil {
		fmt.Printf("%s has nothing to transfer.\n", o.name)
		return
	}
	fmt.Printf("%s transfers %s to %s.\n", o.name, o.report.title, destination.name)
	destination.report = o.report
	o.report = nil
}

func (o *reportOwner) print() {
	if o.report == nil {
		fmt.Printf("%s: empty\n", o.name)
		return
	}
	fmt.Printf("%s: %s\n", o.name, o.report.title)
}

// Fixed inputs make the consequence of using pointers everywhere even when values are simpler
// visible and repeatable.
func main() {
	// These values exercise the normal path before the exercises vary the documented boundaries.
	inbox := reportOwner{name: "Inbox", report: &report{title: "Quarterly Summary"}}
	archive := reportOwner{name: "Archive"}

	inbox.print()
	archive.print()
	inbox.transferTo(&archive)
	inbox.print()
	archive.print()

	score := 90
	primary := &score
	observer := primary
	*observer += 5

	// The printed result shows whether the program can prevent leaks, cycles, and stale observations
	// in ownership graphs.
	fmt.Printf("Shared score after alias update: %d\n", score)
}
